### DISCUSSION TITLE
How to prevent Error: Virtual Users exist, destroy with vudestroy before changing settings

### DISCUSSION THREAD

**Discussion by jsrz** [TPC-Council/HammerDB#622] [Q&A] [answered]
**Guidance**
Bug reports are for when HammerDB is not behaving as expected.
Bug reports should not be submitted for help in understanding database performance related questions.
General questions on database performance or HammerDB usability should be submitted under Discussions.

**Describe the bug**
I'm running TPROC-C against a PostgreSQL DB. I'm using a TCL script to do 6 runs with varying number of virtual user. The first 2-3 runs complete, but then I get errors. Unclear to me how I can prevent this virtual user error. Is something wrong with the script? I've used this same script in the past in older version of HammerDB and PostgreSQL. Not sure what is different.

**Complete output of one of these failed runs:**

`HammerDB CLI v4.8
Copyright (C) 2003-2023 Steve Shaw
Type "help" for a list of commands
Initialized Jobs on-disk database /tmp/hammer.DB using existing tables (7,135,232 KB)
Database set to PostgreSQL

Benchmark set to TPROC-C for PostgreSQL

Value 10.2.108.3 for connection:pg_host is the same as existing value 10.2.108.3, no change made

Value 5432 for connection:pg_port is the same as existing value 5432, no change made

Value disable for connection:pg_sslmode is the same as existing value disable, no change made

Value true for tpcc:pg_timeprofile is the same as existing value true, no change made

Value true for tpcc:pg_storedprocs is the same as existing value true, no change made

Value 1000 for tpcc:pg_count_ware is the same as existing value 1000, no change made

Value 128 for tpcc:pg_num_vu is the same as existing value 128, no change made

Value timed for tpcc:pg_driver is the same as existing value timed, no change made

Value 3 for tpcc:pg_rampup is the same as existing value 3, no change made

Value 15 for tpcc:pg_duration is the same as existing value 15, no change made

Value true for tpcc:pg_raiseerror is the same as existing value true, no change made

Value 1000000000 for tpcc:pg_total_iterations is the same as existing value 1000000000, no change made

Value false for tpcc:pg_connect_pool is the same as existing value false, no change made

Script loaded, Type "print script" to view

Start Test(s)
Testing with 8 users
Vuser 1 created MONITOR - WAIT IDLE

Vuser 2 created - WAIT IDLE

Vuser 3 created - WAIT IDLE

Vuser 4 created - WAIT IDLE

Vuser 5 created - WAIT IDLE

Vuser 6 created - WAIT IDLE

Vuser 7 created - WAIT IDLE

Vuser 8 created - WAIT IDLE

Vuser 9 created - WAIT IDLE

Logging activated
to /tmp/hammerdb_6526EA5E607703E263137343.log

9 Virtual Users Created with Monitor VU

Vuser 1:RUNNING

Vuser 1:Initializing xtprof time profiler

Vuser 1:Beginning rampup time of 3 minutes

Vuser 2:RUNNING

Vuser 2:Initializing xtprof time profiler

Vuser 2:Processing 1000000000 transactions with output suppressed...

Vuser 3:RUNNING

Vuser 3:Initializing xtprof time profiler

Vuser 3:Processing 1000000000 transactions with output suppressed...

Vuser 4:RUNNING

Vuser 4:Initializing xtprof time profiler

Vuser 4:Processing 1000000000 transactions with output suppressed...

Vuser 5:RUNNING

Vuser 5:Initializing xtprof time profiler

Vuser 5:Processing 1000000000 transactions with output suppressed...

Vuser 6:RUNNING

Vuser 6:Initializing xtprof time profiler

Vuser 6:Processing 1000000000 transactions with output suppressed...

Vuser 7:RUNNING

Vuser 7:Initializing xtprof time profiler

Vuser 7:Processing 1000000000 transactions with output suppressed...

Vuser 8:RUNNING

Vuser 8:Initializing xtprof time profiler

Vuser 8:Processing 1000000000 transactions with output suppressed...

Vuser 9:RUNNING

Vuser 9:Initializing xtprof time profiler

Vuser 9:Processing 1000000000 transactions with output suppressed...

Vuser 1:Rampup 1 minutes complete ...

Vuser 1:Rampup 2 minutes complete ...

Vuser 1:Rampup 3 minutes complete ...

Vuser 1:Rampup complete, Taking start Transaction Count.

Vuser 1:Timing test period of 15 in minutes

Vuser 1:1 ...,

Vuser 1:2 ...,

Vuser 1:3 ...,

Vuser 1:4 ...,

Vuser 1:5 ...,

Vuser 1:6 ...,

Vuser 1:7 ...,

Vuser 1:8 ...,

Vuser 1:9 ...,

Vuser 1:10 ...,

Vuser 1:11 ...,

Vuser 1:12 ...,

Vuser 1:13 ...,

Vuser 1:14 ...,

Vuser 1:15 ...,

Vuser 1:Test complete, Taking end Transaction Count.

Vuser 1:8 Active Virtual Users configured

Vuser 1:TEST RESULT : System achieved 196722 NOPM from 453665 PostgreSQL TPM

Vuser 1:Gathering timing data from Active Virtual Users...

Vuser 8:FINISHED SUCCESS

Vuser 3:FINISHED SUCCESS

Vuser 4:FINISHED SUCCESS

Vuser 5:FINISHED SUCCESS

Vuser 9:FINISHED SUCCESS

Vuser 2:FINISHED SUCCESS

Vuser 7:FINISHED SUCCESS

Vuser 6:FINISHED SUCCESS

Vuser 1:Calculating timings...

Vuser 1:Writing timing data to /tmp/hdbxtprofile_6526EE9C607703E293638373.log

Vuser 1:FINISHED SUCCESS

ALL VIRTUAL USERS COMPLETE

vudestroy success

Testing with 16 users
Vuser 1 created MONITOR - WAIT IDLE

Vuser 2 created - WAIT IDLE

Vuser 3 created - WAIT IDLE

Vuser 4 created - WAIT IDLE

Vuser 5 created - WAIT IDLE

Vuser 6 created - WAIT IDLE

Vuser 7 created - WAIT IDLE

Vuser 8 created - WAIT IDLE

Vuser 9 created - WAIT IDLE

Vuser 10 created - WAIT IDLE

Vuser 11 created - WAIT IDLE

Vuser 12 created - WAIT IDLE

Vuser 13 created - WAIT IDLE

Vuser 14 created - WAIT IDLE

Vuser 15 created - WAIT IDLE

Vuser 16 created - WAIT IDLE

Vuser 17 created - WAIT IDLE

Logging activated
to /tmp/hammerdb_6526EEB7607703E213535303.log

17 Virtual Users Created with Monitor VU

Vuser 1:RUNNING

Vuser 1:Initializing xtprof time profiler

Vuser 1:Beginning rampup time of 3 minutes

Vuser 2:RUNNING

Vuser 2:Initializing xtprof time profiler

Vuser 2:Processing 1000000000 transactions with output suppressed...

Vuser 3:RUNNING

Vuser 3:Initializing xtprof time profiler

Vuser 3:Processing 1000000000 transactions with output suppressed...

Vuser 4:RUNNING

Vuser 4:Initializing xtprof time profiler

Vuser 4:Processing 1000000000 transactions with output suppressed...

Vuser 5:RUNNING

Vuser 5:Initializing xtprof time profiler

Vuser 5:Processing 1000000000 transactions with output suppressed...

Vuser 6:RUNNING

Vuser 6:Initializing xtprof time profiler

Vuser 6:Processing 1000000000 transactions with output suppressed...

Vuser 7:RUNNING

Vuser 7:Initializing xtprof time profiler

Vuser 7:Processing 1000000000 transactions with output suppressed...

Vuser 8:RUNNING

Vuser 8:Initializing xtprof time profiler

Vuser 8:Processing 1000000000 transactions with output suppressed...

Vuser 9:RUNNING

Vuser 9:Initializing xtprof time profiler

Vuser 9:Processing 1000000000 transactions with output suppressed...

Vuser 10:RUNNING

Vuser 10:Initializing xtprof time profiler

Vuser 10:Processing 1000000000 transactions with output suppressed...

Vuser 11:RUNNING

Vuser 11:Initializing xtprof time profiler

Vuser 11:Processing 1000000000 transactions with output suppressed...

Vuser 12:RUNNING

Vuser 12:Initializing xtprof time profiler

Vuser 12:Processing 1000000000 transactions with output suppressed...

Vuser 13:RUNNING

Vuser 13:Initializing xtprof time profiler

Vuser 13:Processing 1000000000 transactions with output suppressed...

Vuser 14:RUNNING

Vuser 14:Initializing xtprof time profiler

Vuser 14:Processing 1000000000 transactions with output suppressed...

Vuser 15:RUNNING

Vuser 15:Initializing xtprof time profiler

Vuser 15:Processing 1000000000 transactions with output suppressed...

Vuser 16:RUNNING

Vuser 16:Initializing xtprof time profiler

Vuser 16:Processing 1000000000 transactions with output suppressed...

Vuser 17:RUNNING

Vuser 17:Initializing xtprof time profiler

Vuser 17:Processing 1000000000 transactions with output suppressed...

Vuser 1:Rampup 1 minutes complete ...

Vuser 1:Rampup 2 minutes complete ...

Vuser 1:Rampup 3 minutes complete ...

Vuser 1:Rampup complete, Taking start Transaction Count.

Vuser 1:Timing test period of 15 in minutes

Vuser 1:1 ...,

Vuser 1:2 ...,

Vuser 1:3 ...,

Vuser 1:4 ...,

Vuser 1:5 ...,

Vuser 1:6 ...,

Vuser 1:7 ...,

Vuser 1:8 ...,

Vuser 1:9 ...,

Vuser 1:10 ...,

Vuser 1:11 ...,

Vuser 1:12 ...,

Vuser 1:13 ...,

Vuser 1:14 ...,

Vuser 1:15 ...,

Vuser 1:Test complete, Taking end Transaction Count.

Vuser 1:16 Active Virtual Users configured

Vuser 1:TEST RESULT : System achieved 379697 NOPM from 877211 PostgreSQL TPM

Vuser 1:Gathering timing data from Active Virtual Users...

Vuser 2:FINISHED SUCCESS

Vuser 10:FINISHED SUCCESS

Vuser 6:FINISHED SUCCESS

Vuser 5:FINISHED SUCCESS

Vuser 4:FINISHED SUCCESS

Vuser 11:FINISHED SUCCESS

Vuser 8:FINISHED SUCCESS

Vuser 7:FINISHED SUCCESS

Vuser 12:FINISHED SUCCESS

Vuser 3:FINISHED SUCCESS

Vuser 13:FINISHED SUCCESS

Vuser 16:FINISHED SUCCESS

Vuser 17:FINISHED SUCCESS

Vuser 9:FINISHED SUCCESS

Vuser 1:Calculating timings...

Vuser 14:FINISHED SUCCESS

Vuser 1:Writing timing data to /tmp/hdbxtprofile_6526F2F8607703E253230373.log

Vuser 15:FINISHED SUCCESS

Virtual Users remain running in background or shutting down, retry

Testing with 32 users
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before creating

Virtual Users still active in background
Wait for Virtual Users to finish?

Enter yes or no: replied yes

Error: can't read "threadscreated(0)": no such variable

Virtual Users remain running in background or shutting down, retry

Testing with 64 users
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before creating

Virtual Users still active in background
Wait for Virtual Users to finish?

Enter yes or no: replied yes

Error: can't read "threadscreated(0)": no such variable

Virtual Users remain running in background or shutting down, retry

Testing with 128 users
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before creating

Virtual Users still active in background
Wait for Virtual Users to finish?

Enter yes or no: replied yes

Error: can't read "threadscreated(0)": no such variable

Virtual Users remain running in background or shutting down, retry

Testing with 256 users
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before changing settings
Error: Virtual Users exist, destroy with vudestroy before creating

Virtual Users still active in background
Wait for Virtual Users to finish?

Enter yes or no: replied yes

Error: can't read "threadscreated(0)": no such variable

Virtual Users remain running in background or shutting down, retry

Tests complete`

**To Reproduce**
1. Setup PostgreSQL DB
2. Run buildschema script (see screenshots). This runs ok.
3. Run test script (see screenshots)

**Expected behavior**
6 test runs completed with different numbers of virtual users (8, 16, 32, 64, 128, 256)

**Screenshots**
Build Schema Script:

`#!/bin/tclsh
dbset db pg
dbset bm TPROC-C
diset connection pg_host 10.2.108.3
diset connection pg_port 5432
diset connection pg_sslmode disable
diset tpcc pg_timeprofile true
diset tpcc pg_storedprocs true
diset tpcc pg_count_ware 1000
diset tpcc pg_num_vu 128
diset tpcc pg_driver timed
diset tpcc pg_rampup 3
diset tpcc pg_duration 15
diset tpcc pg_raiseerror true
diset tpcc pg_total_iterations 1000000000
diset tpcc pg_connect_pool false
buildschema
`

Test run script:
`#!/bin/tclsh
dbset db pg
dbset bm TPROC-C
diset connection pg_host 10.2.108.3
diset connection pg_port 5432
diset connection pg_sslmode disable
diset tpcc pg_timeprofile true
diset tpcc pg_storedprocs true
diset tpcc pg_count_ware 1000
diset tpcc pg_num_vu 128
diset tpcc pg_driver timed
diset tpcc pg_rampup 3
diset tpcc pg_duration 15
diset tpcc pg_raiseerror true
diset tpcc pg_total_iterations 1000000000
diset tpcc pg_connect_pool false
loadscript

puts "Start Test(s)"
foreach z { 8 16 32 64 128 256 } {
puts "Testing with $z users"
vuset vu $z
vuset iterations 1
vuset logtotemp 1
vuset unique 1
vucreate
vurun
vudestroy
after 5000
}
puts "Tests complete"
`

**HammerDB Version (please complete the following information):**
 - Version: 4.8
 - Build: Release download of 4.8

**HammerDB Interface (please complete the following information):**
 - UI: CLI

**Operating System (please complete the following information):**
 - Server OS: Ubuntu 23.04
 - Client OS: Ubuntu 23.04

**Database Server (please complete the following information):**
 - Database name: PostgreSQL
 - Database Release Version 16.0

**Database Client (please complete the following information):**
 - Database client name: [e.g. ODBC Driver 18 for SQL Server]

**Additional context**

**Comment by sm-shaw** [comment, marked as answer]
When we close down the Virtual Users after the 16 virtual User run, we get the message:

`Virtual Users remain running in background or shutting down, retry`

If we put the data in order, you can see that Virtual User 1 hasn't finished before we call vudestroy
```
Vuser 1:Gathering timing data from Active Virtual Users...
Vuser 2:FINISHED SUCCESS
Vuser 3:FINISHED SUCCESS
Vuser 4:FINISHED SUCCESS
Vuser 5:FINISHED SUCCESS
Vuser 6:FINISHED SUCCESS
Vuser 7:FINISHED SUCCESS
Vuser 8:FINISHED SUCCESS
Vuser 9:FINISHED SUCCESS
Vuser 10:FINISHED SUCCESS
Vuser 11:FINISHED SUCCESS
Vuser 12:FINISHED SUCCESS
Vuser 13:FINISHED SUCCESS
Vuser 14:FINISHED SUCCESS
Vuser 15:FINISHED SUCCESS
Vuser 16:FINISHED SUCCESS
Vuser 17:FINISHED SUCCESS
Vuser 1:Writing timing data to /tmp/hdbxtprofile_6526F2F8607703E253230373.log
.......
This is not showing --->
.......
Vuser 1:FINISHED SUCCESS
ALL VIRTUAL USERS COMPLETE
vudestroy success
```

So, we need to give it more time and there are a couple of things we can do:

As a quick fix you can put the `after` call before vudestroy, and as for example below this will wait for 60 seconds to let everything finish, you can increase this if its taking longer.
```
after 60000
vudestroy
```
or the correct way in generic.xml - there is a variable called xt_gather_timeout specifically for this purpose. Note that if you can change the XML value you will need to remove the generic.db SQLite database file in your temp area to re-read it. 
```
    <timeprofile>
           <profiler>xtprof</profiler>
           <xt_unique_log_name>0</xt_unique_log_name>
           <xt_gather_timeout>10</xt_gather_timeout>
           <xt_job_storage>1</xt_job_storage>
    </timeprofile>
```
Or for advanced users to set it directly, run a command such as follows either at the CLI prompt or in the console of the HammerDB GUI. 
```
hammerdb>dict set genericdictdb timeprofile xt_gather_timeout 60
theme {scaling auto scaletheme auto pixelsperpoint auto} sqlitedb {sqlitedb_dir TMP} code_highlight {highlight true} benchmark {rdbms MSSQLServer bm TPC-H first_result NOPM} virtual_user_options {virtual_users 1 user_delay 500 repeat_delay 500 iterations 1 show_output 1 log_to_temp 0 unique_log_name 0 no_log_buffer 0 log_timestamps 0} autopilot {apmode disabled autopilot_duration 10 autopilot_sequence {1 2 4 8 12 16 20 24}} mode {hostname localhost id 0} transaction_counter {tc_refresh_rate 10 tc_log_to_temp 0 tc_unique_log_name 0 tc_log_timestamps 0 tc_graph_ribbon false} metrics {agent_hostname localhost agent_id 0} datageneration {gen_count_ware 1 gen_scale_fact 1 gen_num_vu 1} commandline {sqlite_db TMP keepalive_margin 10 jobsoutput JSON jobs_disable 0} webservice {ws_port 8080} timeprofile {profiler xtprof xt_unique_log_name 0 xt_gather_timeout 60 xt_job_storage 1} hdb_version {version v4.9}
```
You can see that xt_gather_timeout is now set to 60, this will then mean that HammerDB will wait for 60 seconds for the writing of the timing data.

**Reply by jsrz** [reply]
First, apologies for opening this as a bug. I didn't realize there was a discussion section. Thanks for making that change.

The removal of generic.db I think is the piece I am missing then. I set xt_gather_timeout to 15 as a test and it didn't resolved the issue. After seeing your response, I went back through the docs (https://www.hammerdb.com/docs/ch01s11.html) and see that it states to not edit the xml files. This is what I was doing when I was on v4.5. Now it makes sense that this method isn't carrying over to v4.8.

I am running in CLI, so I'm going to try your last suggestions. Will mark this as answer when I have a chance to verify.

**Reply by sm-shaw** [reply]
No problem, for a while GitHub didn't have a discussions section at all. 

That makes sense, if it works OK then it could also make sense adding a command to change the generic settings as diset does for the currently selected database XML as diset is doing something similar to the command above underneath.

**Reply by jsrz** [reply]
Ok, this works for me. I've added this command, all is well again.

**Reply by sm-shaw** [reply]
Great, Also to add that the command above will not persist the setting of xt_gather_timeout over restarts. 
To persist the setting also run the command:
`hammerdb>Dict2SQLite "generic" $genericdictdb`
As it sounds this writes the current dict back to the SQLite storage.

**Reply by alfer-edb** [reply]
Hi Steve, hope you are well.

I'm using 4.9 and I'm hitting the same error above. I'm running a simple 8vus test over 50wh to a pg database and when i use xtprof i never get the timing results calculated. I had increased the `after`  to 2 min before `vudestroy` even though the test runtime is just 1 minute but VU1 always results FAILED (for the reasons you explained)

I'll paste the piece of code I use to run the benchmark:

```
#!/bin/tclsh
puts "SETTING CONFIGURATION"
dbset db pg
diset connection pg_host "----"
diset connection pg_port "5432"
diset tpcc pg_raiseerror true
diset tpcc pg_defaultdbase "---"
diset tpcc pg_superuser "-----"
diset tpcc pg_superuserpass "----!"
diset tpcc pg_user "----"
diset tpcc pg_pass "-----!"
diset tpcc pg_count_ware 50
diset tpcc pg_driver timed
diset tpcc pg_duration 1
diset tpcc pg_rampup 1
diset tpcc pg_timeprofile true
diset tpcc pg_allwarehouse true
diset tpcc pg_oracompat true
diset tpcc pg_storedprocs true
vuset logtotemp 1
puts "SEQUENCE STARTED"
foreach z { 8 } {
puts "$z VU TEST"
vuset vu $z
vucreate
vurun
after 120000
vudestroy
        }
puts "TEST SEQUENCE COMPLETE"
exit
```

The resulting output is:

```
...
...
Vuser 9:RUNNING
Vuser 9:Initializing xtprof time profiler
Vuser 9:VU 9 : Assigning WID=8 based on VU count 8, Warehouses = 50 (1 out of 7)
Vuser 9:VU 9 : Assigning WID=16 based on VU count 8, Warehouses = 50 (2 out of 7)
Vuser 9:VU 9 : Assigning WID=24 based on VU count 8, Warehouses = 50 (3 out of 7)
Vuser 9:VU 9 : Assigning WID=32 based on VU count 8, Warehouses = 50 (4 out of 7)
Vuser 9:VU 9 : Assigning WID=40 based on VU count 8, Warehouses = 50 (5 out of 7)
Vuser 9:VU 9 : Assigning WID=48 based on VU count 8, Warehouses = 50 (6 out of 7)
Vuser 9:VU 9 : Assigning WID=6 based on VU count 8, Warehouses = 50 (7 out of 7)
Vuser 9:Processing 10000000 transactions with output suppressed...
Vuser 1:Rampup 1 minutes complete ...
Vuser 1:Rampup complete, Taking start Transaction Count.
Vuser 1:Timing test period of 1 in minutes
Vuser 1:1 ...,
Vuser 1:Test complete, Taking end Transaction Count.
Vuser 1:8 Active Virtual Users configured
Vuser 1:TEST RESULT : System achieved 37 NOPM from 1710 PostgreSQL TPM
Vuser 1:Gathering timing data from Active Virtual Users...
Vuser 4:FINISHED SUCCESS
Vuser 9:FINISHED SUCCESS
Vuser 6:FINISHED SUCCESS
Vuser 8:FINISHED SUCCESS
Vuser 2:FINISHED SUCCESS
Vuser 5:FINISHED SUCCESS
Vuser 3:FINISHED SUCCESS
Vuser 7:FINISHED SUCCESS
Vuser 1:Calculating timings...
Vuser 1:FINISHED FAILED
ALL VIRTUAL USERS COMPLETE
vudestroy success
TEST SEQUENCE COMPLETE
```

Do you see anything wrong ?

**Reply by sm-shaw** [reply]
Alessandro - yes, good to hear from you, 

So yes, there are 2 settings in the XML configuration that control timeouts. The first is xt_gather_timeout for the time for all virtual users to report back their own individual timings to the main thread and the second one is the time for the main thread to wait for the virtual users to finish when running in the command line, by default these are both set to 10.  Being set in the XML configuration, you either need to change this and remove the existing SQLite .db files to pick up the changes or in v4.9 you can go in directly and change the settings manually with the following commands: (example change from 10 seconds to 60 seconds)
```
hammerdb>dict set genericdictdb timeprofile xt_gather_timeout 60
hammerdb>dict set genericdictdb commandline keepalive_margin 60
```
As this is less than intuitive, a change is already coming in the next version to add a "giset" command to modify generic settings and a print generic command here https://github.com/TPC-Council/HammerDB/pull/646

You can use this now by building from source to pikc up all the latest changes which is very easy to do https://www.hammerdb.com/blog/uncategorized/how-to-build-hammerdb-from-source/. Note that you can build just for one database if you wish so you don't need to install the libraries for all the ones you are not interested in just by commenting the unwanted ones out in the Setup file in the build directory.

**Reply by alfer-edb** [reply]
That worked just fine.  I just set the parameters directly into the ingested tcl file and all seems working fine. 
Many thanks!

**Reply by sm-shaw** [reply]
For completeness for anyone reading just the end of the thread: 
From HammerDB v4.10 there is a `giset` command to save generic settings. 
Up to v4.9 to persist the setting over HammerDB restarts also run the command:
`hammerdb>Dict2SQLite "generic" $genericdictdb`
This writes the current generic dict back to the SQLite storage.

**Reply by jsrz** [reply]
Hi @sm-shaw 

So that I understand once I eventually switch to 4.10+

Instead of using a command like
`dict set genericdictdb ...` 

Instead use this `giset` command? Using this command then removes the need to run `Dict2SQLite "generic" $genericdictdb` correct?

**Reply by sm-shaw** [reply]
Yes, that is correct the change is described here: https://github.com/TPC-Council/HammerDB/pull/646
or if you build from the current source then this functionality can be used now as it is already committed.
